File: packages/pluto/src/clients/aws/queue.sns.ts
import { SNSClient, PublishCommand } from "@aws-sdk/client-sns";
import { genResourceId } from "@plutolang/base/utils";
import { CloudEvent, IQueueClient, Queue, QueueOptions } from "../../queue";
import { genAwsResourceName, getAwsAccountId } from "./utils";

/**
 * Implementation of Queue using AWS SNS.
 */
export class SNSQueue implements IQueueClient {
  private readonly id: string;
  private readonly topicArn: Promise<string>;

  private client: SNSClient;

  constructor(name: string, opts?: QueueOptions) {
    this.id = genResourceId(Queue.fqn, name);

    const topicName = genAwsResourceName(this.id);
    this.topicArn = this.buildARN(topicName);
    this.client = new SNSClient({});
    opts;
  }

  public async push(msg: string): Promise<void> {
    const evt: CloudEvent = {
      timestamp: Date.now(),
      data: msg,
    };
    await this.client.send(
      new PublishCommand({
        TopicArn: await this.topicArn,
        Message: JSON.stringify(evt),
      })
    );
  }

  private async buildARN(topicName: string): Promise<string> {
    const region = process.env.AWS_REGION;
    if (!region) {
      throw new Error("Missing AWS Region");
    }

    const accountId = await getAwsAccountId();

    return `arn:aws:sns:${region}:${accountId}:${topicName}`;
  }
}

File: packages/pluto-py/pluto_client/clients/aws/queue_sns.py
import os
import json
import time
import boto3
from typing import Optional
from pluto_base.utils.resource_id import gen_resource_id
from pluto_client.clients.aws.utils import gen_aws_resource_name
from ...queue import CloudEvent, IQueueClient, Queue, QueueOptions


class SNSQueue(IQueueClient):
    def __init__(self, name: str, opts: Optional[QueueOptions] = None) -> None:
        self.__id = gen_resource_id(Queue.fqn, name)
        self.__topic_name = gen_aws_resource_name(self.__id)
        self.__topic_arn = self.__build_arn(self.__topic_name)
        self.__client = boto3.client("sns")

    def push(self, msg: str) -> None:
        event = CloudEvent(timestamp=time.time(), data=msg)
        self.__client.publish(
            TopicArn=self.__topic_arn, Message=json.dumps(event.__dict__)
        )

    def __build_arn(self, topic_name: str) -> str:
        region = os.environ.get("AWS_REGION")
        if not region:
            raise EnvironmentError("Missing AWS Region")

        account_id = os.environ.get("AWS_ACCOUNT_ID")
        if not account_id:
            raise EnvironmentError("Missing AWS Account ID")

        return f"arn:aws:sns:{region}:{account_id}:{topic_name}"
